cannon/mipsevm/exec/mips_instructions.go

```diff
@@ -1,96 +1,114 @@
 }
 
-func ExecMipsCoreStepLogic(cpu *mipsevm.CpuScalars, registers *[32]uint32, memory *memory.Memory, insn, opcode, fun uint32, memTracker MemTracker, stackTracker StackTracker) (memUpdated bool, memAddr uint32, err error) {
+func ExecMipsCoreStepLogic(cpu *mipsevm.CpuScalars, registers *[32]Word, memory *memory.Memory, insn, opcode, fun uint32, memTracker MemTracker, stackTracker StackTracker) (memUpdated bool, memAddr Word, err error) {
 	// j-type j/jal
 	if opcode == 2 || opcode == 3 {
-		linkReg := uint32(0)
+		linkReg := Word(0)
 		if opcode == 3 {
 			linkReg = 31
 		}
 		// Take top 4 bits of the next PC (its 256 MB region), and concatenate with the 26-bit offset
-		target := (cpu.NextPC & 0xF0000000) | ((insn & 0x03FFFFFF) << 2)
+		target := (cpu.NextPC & SignExtend(0xF0000000, 32)) | Word((insn&0x03FFFFFF)<<2)
 		stackTracker.PushStack(cpu.PC, target)
 		err = HandleJump(cpu, registers, linkReg, target)
 		return
 	}
 
 	// register fetch
-	rs := uint32(0) // source register 1 value
-	rt := uint32(0) // source register 2 / temp value
-	rtReg := (insn >> 16) & 0x1F
+	rs := Word(0) // source register 1 value
+	rt := Word(0) // source register 2 / temp value
+	rtReg := Word((insn >> 16) & 0x1F)
 
 	// R-type or I-type (stores rt)
 	rs = registers[(insn>>21)&0x1F]
 	rdReg := rtReg
 	if opcode == 0 || opcode == 0x1c {
 		// R-type (stores rd)
 		rt = registers[rtReg]
-		rdReg = (insn >> 11) & 0x1F
+		rdReg = Word((insn >> 11) & 0x1F)
 	} else if opcode < 0x20 {
 		// rt is SignExtImm
 		// don't sign extend for andi, ori, xori
 		if opcode == 0xC || opcode == 0xD || opcode == 0xe {
 			// ZeroExtImm
-			rt = insn & 0xFFFF
+			rt = Word(insn & 0xFFFF)
 		} else {
 			// SignExtImm
-			rt = SignExtend(insn&0xFFFF, 16)
+			rt = SignExtendImmediate(insn)
 		}
 	} else if opcode >= 0x28 || opcode == 0x22 || opcode == 0x26 {
 		// store rt value with store
 		rt = registers[rtReg]
 
 		// store actual rt with lwl and lwr
 		rdReg = rtReg
+	} else if opcode == 0x27 || opcode == 0x1A || opcode == 0x1B { // 64-bit opcodes
+		// store rt value with store
+		rt = registers[rtReg]
+		// store actual rt with ldl and ldr
+		rdReg = rtReg
 	}
 
 	if (opcode >= 4 && opcode < 8) || opcode == 1 {
 		err = HandleBranch(cpu, registers, opcode, insn, rtReg, rs)
 		return
 	}
 
-	storeAddr := uint32(0xFF_FF_FF_FF)
+	storeAddr := ^Word(0)
 	// memory fetch (all I-type)
 	// we do the load for stores also
-	mem := uint32(0)
+	mem := Word(0)
 	if opcode >= 0x20 {
 		// M[R[rs]+SignExtImm]
-		rs += SignExtend(insn&0xFFFF, 16)
-		addr := rs & 0xFFFFFFFC
+		rs += SignExtendImmediate(insn)
+		addr := rs & arch.AddressMask
 		memTracker.TrackMemAccess(addr)
-		mem = memory.GetMemory(addr)
+		mem = memory.GetWord(addr)
 		if opcode >= 0x28 {
-			// store
-			storeAddr = addr
-			// store opcodes don't write back to a register
-			rdReg = 0
+			// store for 32-bit and certain 64-bit opcodes
+			if arch.IsMips32 || opcode != 0x30 && opcode != 0x34 && opcode != 0x37 {
```

Comment:
`0x30` is for load linked word (`ll`) which is now [handled elsewhere](https://github.com/ethereum-optimism/optimism/blob/dfa61bb2c04d0e89975011c9340cd5efd3b3d78c/cannon/mipsevm/multithreaded/mips.go#L304-L306) and looks like `lld` (0x34) should be handled there as well
```suggestion
			if arch.IsMips32 || opcode != 0x37 {
```